backpack/core/derivatives/lstm.py

```diff
@@ -0,0 +1,351 @@
+            mat: matrix, used to extract device and shapes
+
+        Returns:
+            ifgo, c, c_tanh, h
+        """
+        T: int = mat.shape[1]
+        N: int = mat.shape[2]
+        H: int = module.hidden_size
+        H0: int = 0 * H
+        H1: int = 1 * H
+        H2: int = 2 * H
+        H3: int = 3 * H
+        H4: int = 4 * H
+        # forward pass and save i, f, g, o, c, c_tanh-> ifgo, c, c_tanh
+        ifgo: Tensor = zeros(T, N, 4 * H, device=mat.device, dtype=mat.dtype)
+        c: Tensor = zeros(T, N, H, device=mat.device, dtype=mat.dtype)
+        c_tanh: Tensor = zeros(T, N, H, device=mat.device, dtype=mat.dtype)
+        h: Tensor = zeros(T, N, H, device=mat.device, dtype=mat.dtype)
+        for t in range(T):
+            ifgo[t] = (
+                einsum("hi,ni->nh", module.weight_ih_l0, module.input0[t])
+                + module.bias_ih_l0
+                + module.bias_hh_l0
+            )
+            if t != 0:
+                ifgo[t] += einsum(
+                    "hg,ng->nh", module.weight_hh_l0, module.output[t - 1]
+                )
+            ifgo[t, :, H0:H1] = sigmoid(ifgo[t, :, H0:H1])
+            ifgo[t, :, H1:H2] = sigmoid(ifgo[t, :, H1:H2])
+            ifgo[t, :, H2:H3] = tanh(ifgo[t, :, H2:H3])
+            ifgo[t, :, H3:H4] = sigmoid(ifgo[t, :, H3:H4])
+            c[t] = ifgo[t, :, H0:H1] * ifgo[t, :, H2:H3]
+            if t != 0:
+                c[t] += ifgo[t, :, H1:H2] * c[t - 1]
+            c_tanh[t] = tanh(c[t])
+            h[t] = ifgo[t, :, H3:H4] * c_tanh[t]
+
+        return ifgo, c, c_tanh, h
+
+    @classmethod
+    def _ifgo_jac_t_mat_prod(cls, module: LSTM, mat: Tensor) -> Tensor:
+        V: int = mat.shape[0]
+        T: int = mat.shape[1]
+        N: int = mat.shape[2]
+        H: int = module.hidden_size
+        H0: int = 0 * H
+        H1: int = 1 * H
+        H2: int = 2 * H
+        H3: int = 3 * H
+        H4: int = 4 * H
+
+        ifgo, c, c_tanh, _ = cls._forward_pass(module, mat)
+
+        # backward pass
+        H_prod_t: Tensor = zeros(V, N, H, device=mat.device, dtype=mat.dtype)
+        C_prod_t: Tensor = zeros(V, N, H, device=mat.device, dtype=mat.dtype)
+        C_prod_old: Tensor = zeros(V, N, H, device=mat.device, dtype=mat.dtype)
+        IFGO_prod: Tensor = zeros(V, T, N, 4 * H, device=mat.device, dtype=mat.dtype)
+        for t in reversed(range(T)):
+            # jac_t_mat_prod until node h
+            H_prod_t[:] = mat[:, t]
+            if t != (T - 1):
+                H_prod_t += einsum(
+                    "vnh,hg->vng",
+                    IFGO_prod[:, t + 1],
+                    module.weight_hh_l0,
+                )
+
+            # C_prod_t = jac_t_mat_prod until node c
+            if t != (T - 1):
+                C_prod_old[:] = C_prod_t
+            C_prod_t[:] = einsum(
+                "vnh,nh,nh->vnh",
+                H_prod_t,
+                ifgo[t, :, H3:H4],
+                (1 - c_tanh[t] ** 2),
```

Comment:
Not sure how optimized `einsum` is, but my experience is that if the equation can be simplified easily, this should be done.

Suggestion:
```python
"vnh,nh,nh->vnh",
H_prod_t,
ifgo[t, :, H3:H4] * (1 - c_tanh[t] ** 2),
```

Reply:
Same point for the `einsum` expressions below with similar pattern

Reply:
Oh no! I thought that einsum optimizes automatically... it doesn't...
Yes I will change that.